core/shared/src/main/scala/zio/Schedule2.scala

```diff
@@ -0,0 +1,1009 @@
+/*
+ * Copyright 2017-2020 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio
+
+import java.time.Instant
+import java.util.concurrent.TimeUnit
+
+import zio.duration._
+import zio.random._
+
+/**
+ * A `Schedule[Env, In, Out]` defines a recurring schedule, which consumes values of type `In`, and
+ * which returns values of type `Out`.
+ *
+ * Schedules are defined as a possibly infinite set of intervals spread out over time. Each
+ * interval defines a window in which recurrance is possible.
+ *
+ * When schedules are used to repeat or retry effects, the starting boundary of each interval
+ * produced by a schedule is used as the moment when the effect will be executed again.
+ *
+ * Schedules compose in the following primary ways:
+ *
+ *  * Union. This performs the union of the intervals of two schedules.
```

Comment:
It's trying to take the "earlier" interval when there is no union. That's because if there is no overlap, then the "earlier" interval will determine when the soonest moment the effect can repeat is. Then later, we have a chance for the "later" interval to contribute to the union. Because after the effect is executed due to the first interval, then the schedule will be stepped, and the next interval up may be the "later" interval ("may be" because the schedule with the earlier interval might have another interval before the other one, etc.).

Does that make sense and is the reasoning sound?

Reply:
Yes I think so. It is still union then, in the sense that the overall behaviour of the schedule will likely range over the both intervals in order.